core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java

```diff
@@ -2386,7 +2413,234 @@ protected Scope visitUpdate(Update update, Optional<Scope> scope)
         @Override
         protected Scope visitMerge(Merge merge, Optional<Scope> scope)
         {
-            throw new TrinoException(NOT_SUPPORTED, "This connector does not support merge");
+            Table table = merge.getTable();
+            analysis.setMergeTarget(Optional.of(table));
+            QualifiedObjectName tableName = createQualifiedObjectName(session, table, table.getName());
+            if (metadata.getView(session, tableName).isPresent()) {
+                throw semanticException(NOT_SUPPORTED, merge, "Merging through views is not supported");
+            }
+
+            TableHandle targetTableHandle = metadata.getTableHandle(session, tableName)
+                    .orElseThrow(() -> semanticException(TABLE_NOT_FOUND, table, "Table '%s' does not exist", tableName));
+
+            // Analyzer checks for select permissions but UPDATE has a separate permission, so disable access checks
+            StatementAnalyzer analyzer = new StatementAnalyzer(
+                    analysis,
+                    metadata,
+                    sqlParser,
+                    groupProvider,
+                    new AllowAllAccessControl(),
+                    session,
+                    warningCollector,
+                    CorrelationSupport.ALLOWED);
+
+            Scope targetTableScope = analyzer.analyzeForUpdate(table, scope, UpdateKind.MERGE);
+
+            if (merge.getTargetAlias().isPresent()) {
+                String value = canonicalize(merge.getTargetAlias().get());
+
+                RelationType relationType = targetTableScope.getRelationType();
+                RelationType descriptor = relationType.withAlias(value, null);
+
+                targetTableScope = createAndAssignScope(merge.getTargetAlias().get(), scope, descriptor);
```

Comment:
I'm uncertain of how exactly to do that.  I'll follow up with you or @martint.

Reply:
We discussed this, and I've changed `Merge.target` to be a `Relation` rather than a `Table`, and eliminated `Merge.targetAlias`, instead creating an `AliasedRelation` if the alias is present, and adding `Merge` helper methods to get the underlying `Table` and alias if it exists.